### PR TITLE
Backport linear closures bugfix

### PR DESCRIPTION
**Pull request by jonahbeckford**
This is a backport of https://github.com/ocaml/ocaml/pull/12222 in 5.2.0 to the ("Long Term Support") 4.14 branch.

related: https://discuss.ocaml.org/t/scaling-factors-when-compiling-mutually-recursive-definitions/14708

### CODE REVIEW

**Comment by gasche** [comment]
In principle I think that backporting this is a good idea if users hit this issue -- which is clearly the case here.

@jonahbeckford can you comment on whether any non-trivial rebase work was needed, which would benefit from a review of this own, or if this rebased cleanly?

File: middle_end/closure/closure.ml [hunk 1]
```diff
@@ -882,8 +887,20 @@ let close_approx_var { fenv; cenv } id =
...
   | approx ->
-      let subst = try V.Map.find id cenv with Not_found -> Uvar id in
-      (subst, approx)
+      match cenv with
+      | Not_in_closure -> Uvar id, approx
+      | In_closure { entries; env_param; env_pos } ->
+        let subst =
+          match V.Map.find id entries with
+          | Free_variable fv_pos ->
+            Uprim(P.Pfield (fv_pos - env_pos),
```
Comment (jonahbeckford) [inline middle_end/closure/closure.ml:896]: This line and the following two lines needs a review since it differed from 5.2.0 and caused a merge conflict.

**Comment by jonahbeckford** [comment]
> @jonahbeckford can you comment on whether any non-trivial rebase work was needed, which would benefit from a review of this own, or if this rebased cleanly?

It was not rebased cleanly. I left a comment where there was a merge conflict that should have a review.

**Review by lthls — APPROVED** [review]
I've checked that the small adaptations from the original PR are fine.

**Comment by gasche** [comment]
Merged, thanks!

**Pull request merged**